src/transformers/models/t5/modeling_t5.py

```diff
@@ -434,21 +454,41 @@ def _relative_position_bucket(relative_position, bidirectional=True, num_buckets
         return relative_buckets
 
-    def compute_bias(self, query_length, key_length, device=None):
+    def compute_bias_for_position_embedding_name(self, position_embedding_name, position_ids, query_length, key_length, device=None):
+        """Compute binned relative position bias"""
+        if position_embedding_name == POSITION_EMBEDDING_T5_RELATIVE:
+            relative_attention_bias = self.relative_attention_bias
+            num_buckets=self.relative_attention_num_buckets
+            max_distance=self.relative_attention_max_distance                
+        else:
+            relative_attention_bias = self.relative_attention_bias_dict[position_embedding_name]
+            num_buckets =self.relative_position_embedding_definitions[position_embedding_name].get("num_buckets", self.relative_attention_num_buckets)
+            max_distance=self.relative_position_embedding_definitions[position_embedding_name].get("max_distance", self.relative_attention_max_distance)
+
+        return self.compute_bias(relative_attention_bias=relative_attention_bias, position_ids=position_ids, num_buckets=num_buckets, max_distance=max_distance, query_length=query_length, key_length=key_length, device=device)
+
+    def compute_bias(self, relative_attention_bias, position_ids, num_buckets, max_distance, query_length, key_length, device=None):
         """Compute binned relative position bias"""
         if device is None:
-            device = self.relative_attention_bias.weight.device
-        context_position = torch.arange(query_length, dtype=torch.long, device=device)[:, None]
-        memory_position = torch.arange(key_length, dtype=torch.long, device=device)[None, :]
+            device = relative_attention_bias.weight.device
+        if position_ids is not None:
+            context_position = position_ids.unsqueeze(2)
+            memory_position = position_ids.unsqueeze(1)
+        else:
+            context_position = torch.arange(query_length, dtype=torch.long, device=device)[:, None]
+            memory_position = torch.arange(key_length, dtype=torch.long, device=device)[None, :]
         relative_position = memory_position - context_position  # shape (query_length, key_length)
         relative_position_bucket = self._relative_position_bucket(
             relative_position,  # shape (query_length, key_length)
             bidirectional=(not self.is_decoder),
-            num_buckets=self.relative_attention_num_buckets,
-            max_distance=self.relative_attention_max_distance,
+            num_buckets=num_buckets,
+            max_distance=max_distance,
         )
         values = self.relative_attention_bias(relative_position_bucket)  # shape (query_length, key_length, num_heads)
-        values = values.permute([2, 0, 1]).unsqueeze(0)  # shape (1, num_heads, query_length, key_length)
+        if position_ids is not None:
+            values = values.permute([0, 3, 1, 2])
```

Comment:
Why? can you explain in a comment what each dimension is?

Reply:
values.shape was originally (query_length, key_length, num_heads) - see original comment.
I extended it to account for the case when position_ids are given in input.
So shape is (num_batches - optional, query_length, key_length, num_heads).
Added comments in the code.